File: freight-triage-polyglot-ledger-recovery/environment/intake/src/com/freight/format/Base36Upper.java
package com.freight.format;

/** Formatter: base36 upper. */
public final class Base36Upper implements Formatter {

    private static final String BASE36_DIGITS = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";

    @Override
    public String name() {
        return "base36_upper";
    }

    @Override
    public String apply(long value) {
        if (value == 0L) {
            return "0";
        }
        boolean negative = value < 0L;
        long absolute = negative ? -value : value;
        StringBuilder out = new StringBuilder();
        while (absolute > 0L) {
            out.append(BASE36_DIGITS.charAt((int) (absolute % 36L)));
            absolute /= 36L;
        }
        out.reverse();
        return (negative ? "-" : "") + out;
    }
}

File: freight-triage-polyglot-ledger-recovery/environment/native/src/format/base36_upper.cpp
#include "freight/formats.h"

#include <algorithm>
#include <cstdio>

namespace freight {

// Formatter: base36 upper.
std::string fmt_base36_upper(long long value) {
  static const char kDigits[] = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ";
  if (value == 0) {
    return "0";
  }
  bool negative = value < 0;
  long long absolute = negative ? -value : value;
  std::string out;
  while (absolute > 0) {
    out.push_back(kDigits[absolute % 36]);
    absolute /= 36;
  }
  std::reverse(out.begin(), out.end());
  return negative ? "-" + out : out;
}

}  // namespace freight
